Adds Flutter Template

alberrizwan opened the pull request:
This is draft release of Flutter template.
Please note: 
To manually attach to the device for debugging, follow these steps:

1. SSH into the device and run docker logs <docker container name>.
2. Find the Observatory URL in the logs.
3. Open another SSH terminal and set up port forwarding using this command: ssh -L port:127.0.0.1:port user@deviceIP, where port corresponds to the one in the Observatory URL.
4. Once port forwarding is established, open the Observatory URL in a web browser.
5. You can now start debugging using the web debugger.

Signed-off-by: alber.rizwan@crossware.io

alberrizwan committed:
Adds Flutter Template

andreriesco commented on dartFlutter/Dockerfile.debug:
```diff
@@ -0,0 +1,51 @@
+##
+# Build/Deploy Step
+##
+FROM crossware.app:5050/toradex/flutter:${SDK_BASE_VERSION} as build
```
Hello @alberrizwan,
Sorry for taking a bit of time to the PR review, we are on the process of a new major release of Torizon (Torizon 7).
I created a project form the template and tried to build it, but fell into the following issue:
```
 *  Executing task: DOCKER_HOST= docker compose build --pull --build-arg SSHUSERNAME=torizon --build-arg APP_ROOT=/home/torizon/app --build-arg IMAGE_ARCH=arm64 --build-arg DEBUG_SSH_PORT= --build-arg GPU=-vivante testflutter-debug 

WARN[0000] The "DOCKER_LOGIN" variable is not set. Defaulting to a blank string. 
WARN[0000] /home/andreriesco/Documents/IDE2.0/test-applications/testFlutter/docker-compose.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion 
[+] Building 2.6s (3/3) FINISHED                                        docker-container:sharp_murdock
 => [testflutter-debug internal] load build definition from Dockerfile.debug                      0.3s
 => => transferring dockerfile: 1.23kB                                                            0.0s
 => WARN: FromAsCasing: 'as' and 'FROM' keywords' casing do not match (line 26)                   0.3s
 => WARN: FromAsCasing: 'as' and 'FROM' keywords' casing do not match (line 36)                   0.3s
 => ERROR [testflutter-debug internal] load metadata for crossware.app:5050/toradex/flutter:runt  2.1s
 => CANCELED [testflutter-debug internal] load metadata for crossware.app:5050/toradex/flutter:e  2.1s
------
 > [testflutter-debug internal] load metadata for crossware.app:5050/toradex/flutter:runtime_debug:
------
failed to solve: crossware.app:5050/toradex/flutter:runtime_debug: failed to resolve source metadata for crossware.app:5050/toradex/flutter:runtime_debug: failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://crossware.app/jwt/auth?scope=repository%3Atoradex%2Fflutter%3Apull&service=container_registry: 403 Forbidden

 *  The terminal process "/usr/bin/bash '-c', 'DOCKER_HOST= docker compose build --pull --build-arg SSHUSERNAME=torizon --build-arg APP_ROOT=/home/torizon/app --build-arg IMAGE_ARCH=arm64 --build-arg DEBUG_SSH_PORT= --build-arg GPU=-vivante testflutter-debug'" terminated with exit code: 17. 
 *  Terminal will be reused by tasks, press any key to close it. 
 ```
 It seems that the base container is not publicly available.

andreriesco replied:
Hello @alberrizwan,
Any updates on this issue?

alberrizwan replied:
Hello @andreriesco,
Sorry for the delayed response, I have made the requested changes please check and let us know if any more changes are required.

andreriesco replied:
Hello @alberrizwan ,
Thanks, now I was able to get the base images and run the application.
But I have an issue to debug the application. I followed [the steps do debug it](https://github.com/torizon/vscode-torizon-templates/pull/275#issue-2610731162), but I am falling into the following error:
```
$ dart devtools http://127.0.0.1:43739/k8IUol2dnPI=/
WebSocketException: Connection to 'http://127.0.0.1:43739/k8IUol2dnPI=/ws#' was not upgraded to websocket, HTTP status code: 403
#0      _WebSocketImpl.connect (dart:_http/websocket_impl.dart:1011:41)
#1      WebSocket.connect (dart:_http/websocket.dart:320:22)
#2      vmServiceConnectUriWithFactory (package:vm_service/vm_service_io.dart:26:44)
#3      vmServiceConnectUri (package:vm_service/vm_service_io.dart:12:10)
#4      DevToolsCommand.maybeStartDDS (package:dartdev/src/commands/devtools.dart:192:29)
#5      DevToolsCommand._performDDSCheck (package:dartdev/src/commands/devtools.dart:106:19)
<asynchronous suspension>
#6      DevToolsCommand.run (package:dartdev/src/commands/devtools.dart:57:21)
<asynchronous suspension>
#7      CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>
#8      DartdevRunner.runCommand (package:dartdev/dartdev.dart:240:18)
<asynchronous suspension>
#9      runDartdev (package:dartdev/dartdev.dart:47:16)
<asynchronous suspension>
#10     main (file:///b/s/w/ir/x/w/sdk/pkg/dartdev/bin/dartdev.dart:13:5)
<asynchronous suspension>
```
And even if I try to `disable-service-auth-codes` (and also set a fixed port, as explained in [this article](https://github.com/sony/flutter-embedded-linux/wiki/How-to-debug-Flutter-apps)), I still get an error (but with error code 200):
```
dart devtools http://127.0.0.1:12345/k8IUol2dnPI=/
WebSocketException: Connection to 'http://127.0.0.1:12345/k8IUol2dnPI=/ws#' was not upgraded to websocket, HTTP status code: 200
#0      _WebSocketImpl.connect (dart:_http/websocket_impl.dart:1011:41)
#1      WebSocket.connect (dart:_http/websocket.dart:320:22)
#2      vmServiceConnectUriWithFactory (package:vm_service/vm_service_io.dart:26:44)
#3      vmServiceConnectUri (package:vm_service/vm_service_io.dart:12:10)
#4      DevToolsCommand.maybeStartDDS (package:dartdev/src/commands/devtools.dart:192:29)
#5      DevToolsCommand._performDDSCheck (package:dartdev/src/commands/devtools.dart:106:19)
<asynchronous suspension>
#6      DevToolsCommand.run (package:dartdev/src/commands/devtools.dart:57:21)
<asynchronous suspension>
#7      CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>
#8      DartdevRunner.runCommand (package:dartdev/dartdev.dart:240:18)
<asynchronous suspension>
#9      runDartdev (package:dartdev/dartdev.dart:47:16)
<asynchronous suspension>
#10     main (file:///b/s/w/ir/x/w/sdk/pkg/dartdev/bin/dartdev.dart:13:5)
<asynchronous suspension>
```
Do you know what may be possible issue?

leograba replied:
Hi guys,

Do you have any update about this issue?

alberrizwan replied:
Hi @andreriesco 

I tested the template using Torizon 6, and it ran without issues. However, when I tried reproducing the Issue mentioned with Torizon 7, I encountered an error with the weston-vivante container. The error logs are as follows:

```
[13:04:16.279] failed to initialize display  
[13:04:16.280] EGL error state: EGL_NOT_INITIALIZED (0x3001)  
[13:04:16.280] failed to initialize egl  
00:00:00.191 [ERROR] [seatd/seat.c:332] Could not revoke drm master on device fd: Permission denied  
00:00:00.191 [INFO] [seatd/seat.c:418] No clients on seat0 to activate  
00:00:00.194 [INFO] [seatd/seat.c:524] Closed client 7 on seat0  
00:00:00.194 [INFO] [seatd/seat.c:192] Removed client 7 from seat0  
00:00:00.194 [INFO] [seatd/client.c:471] Client disconnected  
[13:04:16.359] fatal: failed to create compositor backend  
Internal warning: debug scope 'drm-backend' has not been destroyed.  
00:00:00.199 [INFO] [seatd/seatd.c:218] seatd stopped  
Switching back to vt 1
```

Do you have any Suggestions on how to resolve this.

andreriesco replied:
Hello @alberrizwan,
Were you able even to debug the application on Torizon 6? 
For Torizon 7, the major of the base containers need to be changed from `3` to `4`. Also, the `-vivante` containers have been replaced with `-imx8` containers, so for example the `weston-vivante` on Torizon 6 is now `weston-imx8` on Torizon 7

alberrizwan replied:
yes, I was able to debug with torizon 6, and Thanks for the Information on the vivante container, I will make this change and test.

alberrizwan replied:
Hello @andreriesco, could you please confirm if this works with Torizon 6, at first glance the Torizon core 7 seems to have a  significant change especially graphic driver compared to Torizon 6.

andreriesco replied:
Hello @alberrizwan, on Torizon 6 it worked for me, the application is shown on the screen, the only thing that I wasn't able to make work was the debugging [here](https://github.com/torizon/vscode-torizon-templates/pull/275#discussion_r1915665629).
On Torizon 7, some things may have changed on the graphical containers side, but I am not sure which changes could have been made [there](https://github.com/torizon/torizon-containers). But in practice it should not impact the way of using them. 
I would recommend trying to update [this torizon.gpuPrefixRC setting](https://github.com/torizon/vscode-torizon-templates/pull/275/files#diff-e30ea4ad721fdbd5119be98910206f26e384c6fa40c552b44960a2c642838aeeR15) to `true` and maybe trying to update the cgroups, like in [this commit](https://github.com/torizon/vscode-torizon-templates/commit/7f19bda52bf39a81bd2a52fde5ac798f3ff821fa). Also, the name of some containers [had a little change](https://github.com/torizon/vscode-torizon-templates/commit/3fa880baeb51c36ceaafb3efcc0bad045b93f854), but I don't think that you are using any of the ones that changed the name.
Which is the error that is happening to you? Also, which SoM and Carrier Board are you using for testing (right now for example I believe that is a known bug on the native HDMI of the Dahlia carrier board)?

alberrizwan replied:
Hello @andreriesco, Thanks for the Suggestions, I have been able to make it work with torizon 7 by changing the flutter runtime base to wayland-base-imx8 from wayland-base-vivante, however I wasn't to able reproduce the debug websocket Issue on my machines, would it be possible for us to setup a call and Discuss this Issue.

Environment details:
Host device : Ubuntu 22.04
target device: Verdin-iMX8MP

andreriesco commented on dartFlutter/docker-compose.yml:
```diff
@@ -0,0 +1,87 @@
+version: "3.9"
+services:
+  test-debug:
```
Here it needs to be changed to `__container__-debug`

alberrizwan replied:
done

andreriesco commented on dartFlutter/docker-compose.yml:
```diff
@@ -0,0 +1,87 @@
+      weston
+    ]
+
+  test:
```
Here it needs to be changed to `__container__`

alberrizwan replied:
done.

alberrizwan committed:
switch to public registry for base docker containers

alberrizwan committed:
uses public-repos for base containers and makes suggested changes to docker-compose.yml

alberrizwan committed:
Removes redunant files/ folders && fixes template create Issue

alberrizwan committed:
Resolves json error with tasks.json file

alberrizwan committed:
proper formatting for tasks.json file

Commit:
dartFlutter: Fix formatting on JSON files

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
dartFlutter: Update weston major

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
dartFlutter: Minor updates to be formatted with the other templates

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

alberrizwan committed:
Merge branch 'dev' into flutter-7

alberrizwan committed:
Merge pull request #1 from andreriesco/flutter-7

Some small fixes (formatting and weston major)

alberrizwan committed:
Downgrades flutter version for debugging fix

alberrizwan committed:
upgrades flutter version to 3.27.x

alberrizwan committed:
Uses gdb for debugging

Commit:
README: Move README into .doc

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
extensions.json: Add ms-vscode.cpptools extension, for GDB debugging

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
Fix missing new lines

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

alberrizwan committed:
Merge pull request #2 from andreriesco/flutter-7

Small fixes

alberrizwan committed:
Fixes Readme

alberrizwan committed:
adds new line for sanity check

Commit:
tasks: Remove task that comes from common.json

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
tasks: Remove release part on debug

Remove release part on debug, as it is executed by task
run-container-release-<arch>.

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
debugging pipeline: Some updates to make the pipeline faster

Build the debug application saving the build on the workspace, therefore
caching it for the next builds.
Also, copy the application to the board through rsync instead of in a
Docker layer, to avoid the decompression of the application layer on the
board.

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
release: Update release container

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
Fix setting

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
update.json: Update it

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

alberrizwan committed:
Merge pull request #3 from andreriesco/flutter-7

Some updates

alberrizwan committed:
adds flutter install script

alberrizwan committed:
minor fix

alberrizwan committed:
remove mesa from deps

alberrizwan committed:
Fix script Issue

alberrizwan committed:
update Script to more robust

alberrizwan committed:
adds launch.json for local and adds flutter SDk path in settings.json

Commit:
deps.json: Little fix on formatting

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
Little fix on task name

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

Commit:
installFlutter.sh: Add new line in the end

alberrizwan committed:
Merge pull request #4 from andreriesco/flutter-7

Flutter 7

alberrizwan committed:
final changes

alberrizwan committed:
git sanity check

Commit:
dartFlutter: Final updates on documentation

Signed-off-by: Andre Riesco <andre.riesco@toradex.com>

alberrizwan committed:
Merge pull request #5 from andreriesco/flutter-7

dartFlutter: Final updates on documentation

andreriesco commented:
LGTM
